### ISSUE TITLE
Integer overflow vulnerability in vrrp_ipsets_handler 

### ISSUE THREAD

**Issue by destr4ct** [acassen/keepalived#2447]
### Integer overflow vulnerability in vrrp_ipsets_handler

In the vrrp_ipsets_handler handler (file global_parser.c) of the keepalived component, an integer overflow occurs when incorrect arguments are passed. As a result, reading from an undefined address takes place.

```clang
	} else {
		/* No third set specified, copy second name and add "_if6" */
		strcpy_safe(set_name, global_data->vrrp_ipset_address6);
		// [DEBUG] we can specify empty string
                len = strlen(set_name);
		if (set_name[len-1] == '6') // [DEBUG] it's len is zero, so (size_t)(len-1) overflows to 18446744073709551615 and we read from weird index
			set_name[--len] = '\0';
		set_name[IPSET_MAXNAMELEN - 5] = '\0';
		strcat(set_name, "_if6");
		global_data->vrrp_ipset_address_iface6 = STRDUP(set_name);
	}
```

### To Reproduce
1. Configure keepalived with Address Sanitizer (ASAN) and needed libs (check ifdefs)
2. Run keepalived with configuration file specified (keepalived -l -D -n -C -f undermini.txt)

### Expected behavior
The function should handle the case when empty string provided correctly: checking if len is incorrect value (at least not zero)

### Keepalived version
Keepalived v2.3.1 (07/01,2024), git commit v2.3.1-20-gcf0f3e53

### Distros
1. First machine
Name: ubuntu
Version: 22.04
Architecture: aarch64

2. Second machine
Name: ubuntu
Version: 22.04
Architecture: x86-64

### Configuration file
Provided as attachment

### Did keepalived coredump?
No coredump, but ASAN reports a stack-buffer-underflow error.

#### UBSAN message

```
global_parser.c:1137:19: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'size_t' (aka 'unsigned long')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior global_parser.c:1137:19
global_parser.c:1137:7: runtime error: index 18446744073709551615 out of bounds for type 'char[32]'
```

#### ASAN message
```
    #0 0x5555556bcc4f in vrrp_ipsets_handler /home/cha2ned/Desktop/keepalived/keepalived/core/global_parser.c:1137:7
    #1 0x5555558691be in process_stream /home/cha2ned/Desktop/keepalived/lib/parser.c:3098:6
    #2 0x555555869222 in process_stream /home/cha2ned/Desktop/keepalived/lib/parser.c:3103:12
    #3 0x555555866baf in init_data /home/cha2ned/Desktop/keepalived/lib/parser.c:3247:3
    #4 0x5555556a86ec in read_config_file /home/cha2ned/Desktop/keepalived/keepalived/core/main.c:493:2
    #5 0x5555556a86ec in keepalived_main_cfg_harness /home/cha2ned/Desktop/keepalived/keepalived/core/main.c:3089:3
    #6 0x7ffff742a1c9 in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #7 0x7ffff742a28a in __libc_start_main csu/../csu/libc-start.c:360:3
    #8 0x5555555c7084 in _start (/home/cha2ned/Desktop/keepalived/keepalived/keepalived+0x73084) (BuildId: cc10187891197129a48c24a8ebedd956bb130126)
Address 0x7ffff550951f is located in stack of thread T0 at offset 31 in frame
    #0 0x5555556bbd2f in vrrp_ipsets_handler /home/cha2ned/Desktop/keepalived/keepalived/core/global_parser.c:1091

SUMMARY: AddressSanitizer: stack-buffer-underflow /home/cha2ned/Desktop/keepalived/keepalived/core/global_parser.c:1137:7 in vrrp_ipsets_handler
Shadow bytes around the buggy address:
  0x7ffff5509280: f1 f1 f1 f1 f8 f2 f2 f2 f8 f3 f3 f3 00 00 00 00
  0x7ffff5509300: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x7ffff5509380: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x7ffff5509400: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x7ffff5509480: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
=>0x7ffff5509500: f1 f1 f1[f1]00 00 00 00 f3 f3 f3 f3 00 00 00 00
  0x7ffff5509580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffff5509600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffff5509680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffff5509700: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffff5509780: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
```

#### My debug implications
```
 In file: /home/cha2ned/Desktop/astra/projects/tantor/keepalived/keepalived/core/global_parser.c:1137
   1132         }
   1133         else {
   1134                 /* No third set specified, copy second name and add "_if6" */
   1135                 strcpy_safe(set_name, global_data->vrrp_ipset_address6);
   1136                 len = strlen(set_name);
 ► 1137                 if (set_name[len-1] == '6')
   1138                         set_name[--len] = '\0';
   1139                 set_name[IPSET_MAXNAMELEN - 5] = '\0';
   1140                 strcat(set_name, "_if6");
   1141                 global_data->vrrp_ipset_address_iface6 = STRDUP(set_name);
   1142         }

Backtrace
 ► 0   0x5555556bc6fe vrrp_ipsets_handler+2526
   1   0x5555558691bf process_stream+3487
   2   0x555555869223 process_stream+3587
   3   0x555555866bb0 init_data+3760
   4   0x5555556a86ed keepalived_main_cfg_harness+3661
   5   0x5555556a86ed keepalived_main_cfg_harness+3661
   6   0x7ffff742a1ca __libc_start_call_main+122
   7   0x7ffff742a28b __libc_start_main+139
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
pwndbg> p len
$1 = 0
pwndbg> p len-1
$2 = 18446744073709551615
```

**Comment by destr4ct** [comment]
[undermini.txt](https://github.com/user-attachments/files/16176787/undermini.txt)

**Comment by pqarmitage** [comment]
Commit e78513f should resolve this issue. Commits 281de3a and 1e5902c handle empty iptables and nftables chain names respectively.

While reproducing this problem, ASAN identified an error in mem-check debugging when comparing addresses for an rbtree. Commit f3a32e3 resolves that.

@destr4ct:
I have also added a new configure option --enable-sanitize-address. If this is useful for you and you would find it useful to have more related options, please add a comment to this issue.

I have not used ASAN before, so am not familiar with what `and needed libs (check ifdefs)` in your point 1 above. I would be very interested if you could explain that.

**Comment by destr4ct** [comment]
@pqarmitage, thanks for the quick response! Appreciate your offer related to adding more configure options: may you also add
1. option, that turns on the UBSAN
2. option, that appends next cflags "-fprofile-instr-generate -fcoverage-mapping", called somehow like "--enable-cov-collection" (for llvm backend)

**Comment by destr4ct** [comment]
@pqarmitage may I ask if this issue in vrrp_ipsets_handler eligible for gaining CVE record as CWE-190 or something like that?

**Comment by pqarmitage** [comment]
@destr4ct 
I have added the following configure options (in addition to --enable-sanitize-address):
- --enable-sanitize-undefined
- --enable-sanitize-leak
- --enable-sanitize-memory
- --enable-sanitize-hwaddress
- --enable-sanitize-scudo

and also
- --enable-sanitize-address-options
- --enable-sanitize-undefined-options
- --enable-sanitize-leak-options
- --enable-sanitize-memory-options
- --enable-sanitize-hwaddress-options
- --enable-sanitize-scudo-options

which set the default options for each sanitizer.

Not all the sanitize options work with gcc, but they should work with LLVM. I do not have an x86_64 system that supports `--enable-sanitize-hwaddress` but I will test it later on a RasbperryPi (hwaddress is designed for AArch64, but has a workaround to work with some x86_64 processors, if they have a particular feature).

Rather than adding specific support for `-fprofile-instr-generate -fcoverage-mapping` I have added `--enable-cflags`, `--enable-cppflags` and `--enable-ldflags` to be more generic. You should be able to specify `--enable-cflags="-fprofile-instr-generate -fcoverage-mapping"` to achieve what you want.

When looking further at the `vrrp_ipsets` code further I have identified a potential memory leak which commit b1cd430 resolves. There was also an issue where two or more of the configure ipset names could be the same, which would cause an error when adding the ipsets, resolved by commit 7b1cae4. There were similar issues for `vrrp_iptables`, resolved by commit dba70f2. I further discovered that whatever was configured with `vrrp_ipsets`, the configured names were ignored and the default set names used; this is resolved by commit b1accc2.

**Comment by pqarmitage** [comment]
Regarding adding a CVE record, I am not sure it is worthwhile. For the original problem you reported to occur, it required a user to (essentially deliberately since it required use of "") configure an empty ipset name, which is invalid. Further, the ipset names configured would never be used (due to the bug resolved by commit b1accc2), and so I would have hoped that anyone doing so would have raised an issue that configured ipset names were not being used.

I'm not sure if it is valid to look at what a compiler does, or might do, but I have looked at the generated code with GCC 14.1, with optimisation levels -O0, -O1, -O2 and -O3.

First consider the two function local variables - `len` and `set_name`. If len is immediately below set_name on the stack, for the write out of bounds to occur, len must be 0, but since set_name[-1] actually points to a byte in len, it cannot equal '6' (0x36) since len == 0 (gcc -O0).

If len is not immediately below set_name (-O1, -O2 and -O3 where it is in a register), there are two possibilities:
1. set_name[-1] points to below the current stack (-O1 and -O3), in which case any write is to currently unused memory (an alternative is that it points to outside current address space, which should cause a segfault, but since there has been a function call from vrrp_ipsets_handler prior to `if (set_name[-1] == '6')`, then the stack should extend lower).
2. set_name[-1] points to scratch area (-O2) but when `set_name[-1] = '\0'` is executed the top word of the scratch area is unused (this ignores out of order execution but the relevant instructions must be executed in order since they reference the same memory).

With clang (18.1.6), it appears that with -O0 there could be a problem, but only if the most significant byte of strvec is 0x36 and the system is little endian. -O1, -O2 and -O3 seem to fall in the same category as 1. above.

Prior to today's commit adding `--enable-cflags` so far as I can see there was no option to build keepalived other than with -O2.

Having said all the above, if you wish to submit a CVE record then please do so.
